activerecord/lib/active_record/autosave_association.rb

```diff
@@ -505,8 +505,14 @@ def save_belongs_to_association(reflection)
             saved = record.save(validate: !autosave) if record.new_record? || (autosave && record.changed_for_autosave?)
 
             if association.updated?
-              association_id = record.public_send(reflection.options[:primary_key] || :id)
-              self[reflection.foreign_key] = association_id unless self[reflection.foreign_key] == association_id
+              primary_key = Array(reflection.options[:primary_key] || record.class.query_constraints_list)
+              foreign_key = Array(reflection.foreign_key)
+
+              primary_key_foreign_key_pairs = primary_key.zip(foreign_key)
```

Comment:
This breaks unless the `belongs_to` is properly configured. I can see that in the fixtures we have  `Sharded::Comment.belongs_to :blog_post, foreign_key: [:blog_id, :blog_post_id]`, but if you don't, you get an error down the stack.
https://github.com/rails/rails/blob/08f0b74da4fc7eaa7b06fac0bf181e8922f51b0a/activerecord/test/models/sharded/comment.rb#L8

Given the configuration of Sharded::Blog:
https://github.com/rails/rails/blob/08f0b74da4fc7eaa7b06fac0bf181e8922f51b0a/activerecord/test/models/sharded/blog_post.rb#L6

If the foreign key on Comment is not set up properly: `blog_post_id` is mistakenly set to `blog_id`, and we try to set an attribute with an empty name to the `id`.

Since it fails anyway, should we raise if `primary_key` and `foreign_keys` don't have the same size?
We could also fallback to just `:id` as the `primary_key` and use the old behaviour, but I guess long term we want the query constraints to be used in autosave too…

Reply:
>Since it fails anyway, should we raise if primary_key and foreign_keys don't have the same size?

That's a fair option, though I still wanted to explore the possibility of deriving a foreign key like it's done here:
https://github.com/rails/rails/blob/0f5ac79e3e5e716f8699760239a34478b3a0b403/activerecord/lib/active_record/reflection.rb#L720-L727
We already know how to derive `blog_post_id` part of the key as this is the existing convention, so I was wondering if there is a way to tell that `blog_id` is the other part of the foreign key based on `[:blog_id, :id]` association primary key. But if there won't be a convenient way of doing that we should definitely raise if sizes don't match